os/arch/arm/src/imxrt/imxrt_i2s.c

```diff
@@ -0,0 +1,2806 @@
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/* Configuration ************************************************************/
+#define I2S_MAX_BUFFER_SIZE     (4096 - 8)	//the maximum RAM limited by lldesc
+#define PLL_CLK                 (320000000)
+#define I2S_BASE_CLK            (PLL_CLK / 2)
+#define I2S_FULL_DUPLEX_SLAVE_MODE_MASK   (I2S_MODE_TX | I2S_MODE_RX | I2S_MODE_SLAVE)
+#define I2S_FULL_DUPLEX_MASTER_MODE_MASK  (I2S_MODE_TX | I2S_MODE_RX | I2S_MODE_MASTER)
+
+#define APLL_MIN_FREQ (250000000)
+#define APLL_MAX_FREQ (500000000)
+#define APLL_I2S_MIN_RATE (10675)	//in Hz, I2S Clock rate limited by hardware
+
+#ifndef CONFIG_SCHED_WORKQUEUE
+#error Work queue support is required (CONFIG_SCHED_WORKQUEUE)
+#endif
+
+#ifndef CONFIG_AUDIO
+#error CONFIG_AUDIO required by this driver
+#endif
+
+#ifndef CONFIG_IMXRT_I2S_MAXINFLIGHT
+#define CONFIG_IMXRT_I2S_MAXINFLIGHT 32
+#endif
+
+/* Assume no RX/TX support until we learn better */
+#undef I2S_HAVE_RX
+#undef I2S_HAVE_TX
+
+#if defined(CONFIG_IMXRT_I2S)
+
+/* Check for I2S RX support */
+#if defined(CONFIG_IMXRT_I2S_RX)
+#define I2S_HAVE_RX     1
+#endif
+
+/* Check for I2S TX support */
+#if defined(CONFIG_IMXRT_I2S_TX)
+#define I2S_HAVE_TX     1
+#endif
+
+#ifndef CONFIG_IMXRT_I2S_DATALEN
+#define CONFIG_IMXRT_I2S_DATALEN        (I2S_BITS_PER_SAMPLE_16BIT)
+#endif
+
+#ifndef CONFIG_IMXRT_I2S_SAMPLERATE
+#define CONFIG_IMXRT_I2S_SAMPLERATE     (44100)
+#endif
+
+#ifndef CONFIG_IMXRT_I2S_CHANNEL_FMT
+#define CONFIG_IMXRT_I2S_CHANNEL_FMT    (I2S_CHANNEL_FMT_RIGHT_LEFT)
+#endif
+
+#ifndef CONFIG_IMXRT_I2S_COMM_FMT
+#define CONFIG_IMXRT_I2S_COMM_FMT       I2S_COMM_FORMAT_I2S
+#endif
+
+#ifndef CONFIG_IMXRT_I2S_INTR_LEVEL
+#define CONFIG_IMXRT_I2S_INTR_LEVEL    (1)
+#endif
+
+#endif
+
+#ifdef CONFIG_DEBUG
+#define CONFIG_DEBUG_FEATURES
+#endif
+
+#define DEBUG_I2S_DRIVER    0
+
+#undef i2serr
+#undef i2sinfo
+#if defined(DEBUG_I2S_DRIVER) && (0 < DEBUG_I2S_DRIVER)
+#define i2sinfo(format, ...)   printf(format, ##__VA_ARGS__)
+#define i2serr(format, ...)   printf(format, ##__VA_ARGS__)
+#else
+#define i2sinfo(format, ...)
+//#define i2serr(format, ...)
+#endif
+#define i2serr(format, ...)   printf(format, ##__VA_ARGS__)
+
+#if defined(I2S_HAVE_RX) || defined(I2S_HAVE_TX)
+
+#ifndef MIN
+#define MIN(a,b) a < b ? a : b
+#endif
+
+#define OVER_SAMPLE_RATE (384U)
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+enum _sai_transfer_state
+{
+    kSAI_Busy = 0x0U, /*!< SAI is busy */
+    kSAI_Idle,        /*!< Transfer is done. */
+    kSAI_Error        /*!< Transfer error occured. */
+};
+
+/* I2S configuration parameters for i2s_param_config function */
+typedef struct {
+	i2s_mode_t mode;			/*!< I2S work mode */
+	int sample_rate;			/*!< I2S sample rate */
```

Comment:
please align tinyalsa audio system.
we using unsigned int for channel, samplerate, framesize.

Reply:
I have referred to the source in esp32_12s.c file
How can I check it with your sources about the comments that you said to me?
In other words, I want to check it in your tinyalsa source before I change my source